### PR TITLE
ET-354-add-allergy-action-to-elation

### PR DESCRIPTION
**Pull request by sharlotta93**
### **User description**
Add new allergy action for Elation

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Implemented the `createAllergy` action to add allergies to patient profiles, including comprehensive error handling for validation and server errors.
- Added mock functions and constants to support allergy creation in tests.
- Developed unit tests for the `createAllergy` action to ensure correct functionality and error handling.
- Updated client classes to include methods for creating allergies.
- Defined necessary types and validation schemas for allergy data.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>client.ts</strong><dd><code>Add mock function for allergy creation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/__mocks__/client.ts

<li>Added a mock function <code>createAllergy</code> to simulate allergy creation.<br> <li> Included <code>allergyExample</code> in the imports.<br>

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/505/files#diff-f233b91df2b9bf19f5f13ccc7522c18dd4a8085ebdb2812dc8dfab8a04372049">+7/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>constants.ts</strong><dd><code>Define allergy example constant</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/__mocks__/constants.ts

- Added `allergyExample` constant with sample allergy data.

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/505/files#diff-1620308b774afef7cdd55c47a2fc1c5d728f51202540645730dc217894556fb6">+8/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>createAllergy.ts</strong><dd><code>Implement createAllergy action with error handling</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/actions/createAllergy.ts

<li>Implemented <code>createAllergy</code> action for adding allergies to patient <br>profiles.<br> <li> Included error handling for validation and server errors.<br>

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/505/files#diff-222e7861a98010b4fbc23b162b6d96967639f5e80292657a939090850bbc88df">+143/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Register createAllergy action</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/actions/index.ts

- Added `createAllergy` to the list of available actions.

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/505/files#diff-9a1c0055b212081498d2441b821c3c51501a8ff23fa020ebccd3e99395c37a08">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>client.ts</strong><dd><code>Add createAllergy method to client classes</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/client.ts

<li>Added <code>createAllergy</code> method to <code>ElationDataWrapper</code> and <code>ElationAPIClient</code>.<br> <br>

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/505/files#diff-65c140fbe1debe73b11455b75ab0aa6aa573b3c47babb49245c0d35c8e62a662">+19/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>allergy.ts</strong><dd><code>Define types for allergy creation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/types/allergy.ts

<li>Defined types for <code>CreateAllergyInput</code> and <code>CreateAllergyResponse</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/505/files#diff-1f53dbe7f44433232b087978282dd2828ef3ea19caf22036ebab0a8dd57666da">+8/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>allergy.zod.ts</strong><dd><code>Define Zod schema for allergy validation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/validation/allergy.zod.ts

- Created `allergySchema` for validating allergy data.

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/505/files#diff-b64e36f722a464b5efa8bc52ee703274941d5f3e204a147536d3c7b913e30679">+12/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>createAllergy.ts</strong><dd><code>Add tests for createAllergy action</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/actions/__tests__/createAllergy.ts

<li>Added tests for <code>createAllergy</code> action.<br> <li> Included test cases for successful creation and error handling.<br>

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/505/files#diff-8983a209ca7d267d7c06e048a3ced55476588b6219bd22da04b8226bc05518e5">+97/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: extensions/elation/actions/createAllergy.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+} satisfies Record<string, DataPointDefinition>
+
+export const createAllergy: Action<
+  typeof fields,
+  typeof settings,
+  keyof typeof dataPoints
+> = {
+  key: 'createAllergy',
+  category: Category.EHR_INTEGRATIONS,
+  title: 'Add Allergy',
```
Comment (nckhell) [inline extensions/elation/actions/createAllergy.ts:69]: Inconsistency between "Create allergy" and "Add allergy"

File: extensions/elation/actions/createAllergy.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  description:
+    'Adds allergy to the patient profile. If allergy.name is NKDA, we will document a structured NKDA in the patient chart instead of creating a patient allergy called "NKDA"',
```
Comment (nckhell) [inline extensions/elation/actions/createAllergy.ts:71]: How does this happen? I don't see any logic in the code that does this?

File: extensions/elation/actions/createAllergy.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    try {
+      const { patientId, name, startDate, reaction, severity } = payload.fields
+
+      const allergy = allergySchema.parse({
+        patient: patientId,
+        name,
+        start_date: startDate,
+        reaction,
+        severity,
+      })
```
Comment (nckhell) [inline extensions/elation/actions/createAllergy.ts:85]: From experience, I recommend decoupling the action field schema from the API schema. In some cases, they can be identical but in many other cases they can deviate.

File: extensions/elation/actions/createAllergy.ts [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            {
+              date: new Date().toISOString(),
+              text: { en: error.message },
+              error: {
+                category: 'WRONG_INPUT',
+                message: error.message,
+              },
+            },
+          ],
+        })
```
Comment (nckhell) [inline extensions/elation/actions/createAllergy.ts:108]: Not needed, this is caught on the extension server.

File: extensions/elation/actions/createAllergy.ts [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            {
+              date: new Date().toISOString(),
+              text: { en: message },
+              error: {
+                category: 'SERVER_ERROR',
+                message,
+              },
+            },
+          ],
+        })
```
Comment (nckhell) [inline extensions/elation/actions/createAllergy.ts:139]: Not needed. This caught on the extension server.

File: extensions/elation/types/allergy.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import { type z } from 'zod'
+import { type allergySchema } from '../validation/allergy.zod'
```
Comment (nckhell) [inline extensions/elation/types/allergy.ts:2]: We usually keep this in the same file. 

Example: https://github.com/awell-health/awell-extensions/blob/ET-354-add-allergy-action-to-elation/extensions/bland/api/schema/SendCall.schema.ts

**Review by nckhell — COMMENTED** [review]
Nitpick and optional: we usually stick to the following folder structure for extension actions:

```
/extensions
   /extensionKey
      /actions
         /actionKey
            /config
               fields.ts
               datapoints.ts
            actionKey.ts
            actionKey.test.ts
```

I.e. every action has its dedicated folder and inputs (fields) and outputs (datapoints) are stored in a separate file.

File: extensions/elation/actions/createAllergy.ts (hunk 2 above)
Comment (sharlotta93) [reply extensions/elation/actions/createAllergy.ts]: this is copy paste from their website so this happens on Elation side - I've added it here as a warning to the user ( I highly doubts users are in a habit of going to elation docs to check things so it made sense to add it here

File: extensions/elation/actions/createAllergy.ts (hunk 1 above)
Comment (sharlotta93) [reply extensions/elation/actions/createAllergy.ts]: I wanted to keep the lingo in the code same as elation documentation - but to the user it's weird to have 'create allergy' hence I opted for that.

**Pull request merged**